File: example/low_level_motion_example/low_level_motion_example.cpp
#include "magic_robot.h"

#include <unistd.h>
#include <csignal>

#include <iostream>

using namespace magic::gen1;

magic::gen1::MagicRobot robot;

void signalHandler(int signum) {
  std::cout << "Interrupt signal (" << signum << ") received.\n";

  robot.Shutdown();
  // 退出进程
  exit(signum);
}

int main() {
  // 绑定 SIGINT（Ctrl+C）
  signal(SIGINT, signalHandler);

  std::string local_ip = "192.168.54.111";
  // 配置本机网线直连机器的IP地址，并进行SDK初始化
  if (!robot.Initialize(local_ip)) {
    std::cerr << "robot sdk initialize failed." << std::endl;
    robot.Shutdown();
    return -1;
  }

  // 连接机器人
  auto status = robot.Connect();
  if (status.code != ErrorCode::OK) {
    std::cerr << "connect robot failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    robot.Shutdown();
    return -1;
  }

  // 切换运控控制器为底层控制器，默认是高层控制器
  status = robot.SetMotionControlLevel(ControllerLevel::LowLevel);
  if (status.code != ErrorCode::OK) {
    std::cerr << "switch robot motion control level failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    robot.Shutdown();
    return -1;
  }

  // 获取底层控制器
  auto& controller = robot.GetLowLevelMotionController();

  // 设置控制指令发送周期为2ms，500HZ
  controller.SetPeriodMs(2);

  // 订阅imu数据
  controller.SubscribeBodyImu([](const std::shared_ptr<Imu> msg) {
    static int32_t count = 0;
    if (count++ % 1000 == 1) {
      std::cout << "receive imu data." << std::endl;
    }
    // TODO: handle imu data
  });

  // 订阅手部数据
  controller.SubscribeArmState([](const std::shared_ptr<JointState> msg) {
    static int32_t count = 0;
    if (count++ % 1000 == 1) {
      std::cout << "receive arm joint data." << std::endl;
    }
    // TODO: handle arm joint data
  });

  // 以上臂关节控制为例：

  // 第一次发布手部控制指令，关节的操作模式为200，表示关节处于准备状态
  {
    JointCommand arm_command;
    arm_command.joints.resize(kArmJointNum);
    for (int ii = 0; ii < kArmJointNum; ii++) {
      arm_command.joints[ii].operation_mode = 200;
    }
    controller.PublishArmCommand(arm_command);
  }

  // 后续关节控制指令，关节的操作模式为1，表示关节处于位置控制模式
  {
    // 左臂关节，参考文档：
    // 左臂或者右臂1-5关节operation_mode需要从模式：200切换到模式：4（串联PID模式）进行指令下发；
    JointCommand arm_command;
    arm_command.joints.resize(kArmJointNum);
    for (int ii = 0; ii < 5; ii++) {
      arm_command.joints[ii].operation_mode = 4;
      // TODO:设置目标位置、速度、力矩和增益
    }

    for (int ii = 5; ii < 6; ii++) {
      arm_command.joints[ii].operation_mode = 5;
      // TODO:设置目标位置、速度、力矩和增益
    }

    for (int ii = 7; ii < 12; ii++) {
      arm_command.joints[ii].operation_mode = 4;
      // TODO:设置目标位置、速度、力矩和增益
    }

    for (int ii = 12; ii < 13; ii++) {
      arm_command.joints[ii].operation_mode = 5;
      // TODO:设置目标位置、速度、力矩和增益
    }
    controller.PublishArmCommand(arm_command);
  }

  usleep(100000000);

  // 断开与机器人的链接
  status = robot.Disconnect();
  if (status.code != ErrorCode::OK) {
    std::cerr << "disconnect robot failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    robot.Shutdown();
    return -1;
  }

  robot.Shutdown();

  return 0;
}

File: example/python/low_level_motion_example.py
#!/usr/bin/env python3

import sys
import time
import signal
import logging
from typing import Optional

import magicbot_gen1_python as magicbot

# Configure logging format and level
logging.basicConfig(
    level=logging.INFO,  # Minimum log level
    format="%(asctime)s [%(levelname)s] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)


# Global variables
robot: Optional[magicbot.MagicRobot] = None


def signal_handler(signum, frame):
    """Signal handler function for graceful exit"""
    global robot
    logging.info("Received interrupt signal (%s), exiting...", signum)
    if robot:
        robot.disconnect()
        robot.shutdown()


def body_imu_callback(imu_data):
    """Body IMU data callback function"""
    logging.info("Received body IMU data")
    # TODO: Process body IMU data


def arm_state_callback(joint_state):
    """Arm joint state callback function"""
    logging.info("Received arm joint state data")
    # TODO: Process arm joint state data


def main():
    """Main function"""
    global robot

    # Bind signal handler
    signal.signal(signal.SIGINT, signal_handler)

    logging.info("Robot model: %s", magicbot.get_robot_model())

    # Create robot instance
    robot = magicbot.MagicRobot()

    try:
        # Configure local IP address for direct network connection and initialize SDK
        local_ip = "192.168.54.111"
        if not robot.initialize(local_ip):
            logging.error("Failed to initialize robot SDK")
            robot.shutdown()
            return -1

        # Connect to robot
        status = robot.connect()
        if status.code != magicbot.ErrorCode.OK:
            logging.error(
                "Failed to connect to robot, code: %s, message: %s",
                status.code,
                status.message,
            )
            robot.shutdown()
            return -1

        logging.info("Successfully connected to robot")

        # Switch motion control controller to low-level controller, default is high-level controller
        status = robot.set_motion_control_level(magicbot.ControllerLevel.LowLevel)
        if status.code != magicbot.ErrorCode.OK:
            logging.error(
                "Failed to switch robot motion control level, code: %s, message: %s",
                status.code,
                status.message,
            )
            robot.shutdown()
            return -1

        logging.info("Switched to low-level motion controller")

        # Get low-level motion controller
        controller = robot.get_low_level_motion_controller()

        # Set control command sending period to 2ms, 500Hz
        controller.set_period_ms(2)
        logging.info("Set control period to 2ms (500Hz)")

        # Subscribe to body IMU data
        controller.subscribe_body_imu(body_imu_callback)
        logging.info("Subscribed to body IMU data")

        # Subscribe to arm joint state
        controller.subscribe_arm_state(arm_state_callback)
        logging.info("Subscribed to arm joint state")

        # Create arm joint control command
        arm_command = magicbot.JointCommand()
        arm_command.joints = []

        # Set all joints to preparation state (operation mode 200)
        for i in range(magicbot.ARM_JOINT_NUM):
            joint = magicbot.SingleJointCommand()
            joint.operation_mode = 200  # Preparation state
            joint.pos = 0.0
            joint.vel = 0.0
            joint.toq = 0.0
            joint.kp = 0.0
            joint.kd = 0.0
            arm_command.joints.append(joint)

        # Publish arm joint control command
        controller.publish_arm_command(arm_command)
        logging.info(
            "Published arm joint preparation state command, joint count: %d",
            len(arm_command.joints),
        )

        # Wait 100 seconds
        logging.info("Waiting 100 seconds...")
        time.sleep(100)

        return 0

    except Exception as e:
        logging.error("Exception occurred during program execution: %s", e)
        return -1

    finally:
        # Clean up resources
        try:
            # Close low-level motion controller
            controller = robot.get_low_level_motion_controller()
            controller.shutdown()
            logging.info("Low-level motion controller closed")

            # Disconnect
            robot.disconnect()
            logging.info("Robot connection disconnected")

            # Shutdown robot
            robot.shutdown()
            logging.info("Robot shutdown")

        except Exception as e:
            logging.error("Exception occurred while cleaning up resources: %s", e)


if __name__ == "__main__":
    sys.exit(main())
